File: services/api/schemas/exit_proximity.py
"""services/api/schemas/exit_proximity.py — response models for /api/today/exit-proximity.

Crypto-pivot §3.9 (``Docs/crypto-pivot-delta-spec.md``): the exit-side
"Stops" projection. The pre-pivot CME exit-trigger table (stop /
trend-flip / reversal / decommission chips off the ``exit_proximity``
table, ``Docs/exit-proximity-design.md``) died with the LEAN cycle;
exit proximity is now **distance to the client 2xATR stop and the
native 3xATR venue backstop** per open position, composed from
``positions_current`` (+ ``contracts`` enrichment), the strategy
worker's ``engine_state`` stops, the latest native-stop ``orders`` row,
and the api-resident live mark store (``services/api/marks``).

The URL is preserved; the response SHAPE is wire-breaking by design —
the §3.9 PR swaps both sides atomically.

**Headroom convention** (server-computed; the frontend only formats):
``(mark - stop) / mark`` for a long, ``(stop - mark) / mark`` for a
short — a positive fraction is distance-to-stop remaining; None when
either the mark or the stop is missing (no honest number to show).
Decimal throughout per [A05]; quantized to 6 dp for a deterministic
wire string.

**Ordering**: closest-to-stop first (smallest non-null headroom across
the two stops), with stopped-today / unprotected rows (non-zero
contracts but no native stop resting at the venue) sorted to the very
top — the rows the operator must see first.

Every mapper below is a pure function so the unit tests pin the math
(``tests/unit/test_api_exit_proximity_route.py``) — no clock reads,
no I/O.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Final, Literal

from pydantic import BaseModel, ConfigDict, Field

DirectionLiteral = Literal["long", "short"]

#: Wire-determinism quantum for the headroom fractions (Decimal division
#: otherwise widens the scale unpredictably).
_HEADROOM_QUANTUM: Final[Decimal] = Decimal("0.000001")


class PositionStopProximityView(BaseModel):
    """One open position's stop-proximity row (delta spec §3.9)."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    product_id: str = Field(description="CDE product id (positions_current.market).")
    asset: str | None = Field(
        default=None,
        description=(
            "Base-asset label (BTC/ETH) from contracts.market_root; None "
            "when the row has no contracts enrichment."
        ),
    )
    direction: DirectionLiteral
    contracts: int = Field(description="Signed net contracts held in this product.")
    entry_vwap: Decimal | None = Field(
        default=None,
        description="Quantity-weighted average entry price (positions_current.avg_cost).",
    )
    mark: Decimal | None = Field(
        default=None,
        description=(
            "Live WS mark from the api-resident mark store; None when the "
            "store is absent or the mark exceeds the 3-minute staleness "
            "policy (services/api/marks)."
        ),
    )
    marks_stale: bool = Field(
        description=(
            "The strategy worker's marks_stale flag; True (conservative) "
            "when the worker has never reported."
        ),
    )
    client_stop_price: Decimal | None = Field(
        default=None,
        description=(
            "The worker's client-side 2xATR soft stop "
            "(engine_state.client_stop_level); enforced by the 30 s risk loop."
        ),
    )
    client_stop_headroom_frac: Decimal | None = Field(
        default=None,
        description="Distance-to-client-stop fraction (module docstring convention).",
    )
    native_stop_price: Decimal | None = Field(
        default=None,
        description=(
            "Resting venue 3xATR stop-limit backstop price (latest orders "
            "row for the worker's native_stop_client_order_id)."
        ),
    )
    native_stop_headroom_frac: Decimal | None = Field(
        default=None,
        description="Distance-to-native-stop fraction (module docstring convention).",
    )
    native_stop_resting: bool = Field(
        description=(
            "True when the worker tracks a venue-resting native stop order "
            "id for this asset. False with non-zero contracts = UNPROTECTED."
        ),
    )
    stopped_today: bool = Field(
        description="True when the engine's stop fired on today's UTC bar-day.",
    )


class ExitProximityResponse(BaseModel):
    """``GET /api/today/exit-proximity`` envelope.

    Fresh-DB / no-open-positions contract: ``positions`` is empty —
    HTTP 200, never an error.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    positions: list[PositionStopProximityView]
    server_now: datetime


# ---------------------------------------------------------------------------
# Pure mappers (math pinned by tests)
# ---------------------------------------------------------------------------


def stop_headroom_frac(
    *,
    mark: Decimal | None,
    stop: Decimal | None,
    direction: DirectionLiteral,
) -> Decimal | None:
    """Distance-to-stop as a fraction of the mark (module docstring).

    ``(mark - stop)/mark`` for longs, ``(stop - mark)/mark`` for shorts;
    None when either input is missing or the mark is non-positive (no
    honest denominator).
    """
    if mark is None or stop is None or mark <= 0:
        return None
    raw = (mark - stop) / mark if direction == "long" else (stop - mark) / mark
    return raw.quantize(_HEADROOM_QUANTUM)


def build_position_stop_view(
    *,
    product_id: str,
    asset: str | None,
    contracts: int,
    entry_vwap: Decimal | None,
    mark: Decimal | None,
    marks_stale: bool,
    client_stop_price: Decimal | None,
    native_stop_price: Decimal | None,
    native_stop_resting: bool,
    stopped_today: bool,
) -> PositionStopProximityView:
    """Compose one row: direction from the sign of ``contracts`` + the two
    server-computed headroom fractions. ``contracts`` must be non-zero
    (flat rows are filtered by the caller)."""
    direction: DirectionLiteral = "long" if contracts > 0 else "short"
    return PositionStopProximityView(
        product_id=product_id,
        asset=asset,
        direction=direction,
        contracts=contracts,
        entry_vwap=entry_vwap,
        mark=mark,
        marks_stale=marks_stale,
        client_stop_price=client_stop_price,
        client_stop_headroom_frac=stop_headroom_frac(
            mark=mark, stop=client_stop_price, direction=direction
        ),
        native_stop_price=native_stop_price,
        native_stop_headroom_frac=stop_headroom_frac(
            mark=mark, stop=native_stop_price, direction=direction
        ),
        native_stop_resting=native_stop_resting,
        stopped_today=stopped_today,
    )


def _sort_key(view: PositionStopProximityView) -> tuple[int, Decimal, str]:
    """Closest-to-stop first; stopped/unprotected rows before everything.

    Primary: attention group (stopped today OR no native stop resting =
    0, else 1). Secondary: MIN non-null headroom ascending (rows with no
    computable headroom sort last within their group via +Infinity).
    Tertiary: product_id for a stable, deterministic order.
    """
    attention = 0 if (view.stopped_today or not view.native_stop_resting) else 1
    headrooms = [
        h for h in (view.client_stop_headroom_frac, view.native_stop_headroom_frac) if h is not None
    ]
    min_headroom = min(headrooms) if headrooms else Decimal("Infinity")
    return (attention, min_headroom, view.product_id)


def sort_stop_views(
    views: list[PositionStopProximityView],
) -> list[PositionStopProximityView]:
    """Return a NEW list ordered per :func:`_sort_key` (input unmutated)."""
    return sorted(views, key=_sort_key)


__all__ = [
    "DirectionLiteral",
    "ExitProximityResponse",
    "PositionStopProximityView",
    "build_position_stop_view",
    "sort_stop_views",
    "stop_headroom_frac",
]

File: apps/web/src/lib/api/exit-proximity.ts
/**
 * apps/web/src/lib/api/exit-proximity.ts — typed client for
 * GET /api/today/exit-proximity (the Today page's "Stops" view).
 *
 * Crypto-pivot §3.9 rewrite: the exit-side view is now per-open-position
 * STOP proximity — distance to the client 2×ATR soft stop (enforced by
 * the strategy worker's 30 s risk loop) and the native 3×ATR venue
 * backstop — not the retired CME exit-trigger table. Wire contract
 * mirrors `services/api/schemas/exit_proximity.py` (Pydantic v2
 * `BaseModel` with `extra='forbid'`) FIELD-BY-FIELD.
 *
 * Decimal-on-wire convention (A05 / dev-guide §3.8): every numeric field
 * (`entry_vwap`, `mark`, stop prices, headroom fractions) arrives as a
 * Decimal-as-string. Per frontend-spec §8.9 (LOCKED) we DO NOT do
 * Decimal arithmetic in JS — the headroom fractions are computed
 * SERVER-SIDE; this file only classifies rows on booleans and formats
 * the provided strings at the display layer (`parseFloat` on bounded
 * fractional values only).
 *
 * ⚠️ TWO THINGS CARRIED OVER FROM THE PRE-PIVOT CLIENT — read before editing:
 *
 *   1. NO CLIENT-SIDE SORT. The endpoint returns `positions` ALREADY
 *      ordered closest-to-stop (stopped/unprotected rows first, then
 *      ascending min headroom, server-side). We render in RECEIPT ORDER.
 *
 *   2. INVERTED COLOR VALENCE. green = HOLDING = "position SAFE",
 *      red = TRIGGERED = "stop fired / unprotected" — the OPPOSITE
 *      valence of the entry side. We keep a SEPARATE state→token map
 *      (`exitStateChipClass`) so a shared chip component can never paint
 *      a stopped position green.
 *
 * Refresh model: TanStack Query with `staleTime`, keyed under
 * `['positions', 'exit-proximity']` so the existing `position` AND `fill`
 * SSE event handlers in `lib/sse.ts` (both prefix-invalidate
 * `['positions']`) also refresh this view. No new SSE event type is
 * introduced (forbidden by CLAUDE.md locked rules).
 */

import { useQuery, type UseQueryResult } from '@tanstack/react-query';

import { apiCall } from '../api';

// ---------------------------------------------------------------------------
// Wire types (1:1 with services/api/schemas/exit_proximity.py)
// ---------------------------------------------------------------------------

export type DirectionLiteral = 'long' | 'short';

export interface PositionStopProximityView {
  /** CDE product id (e.g. `BIP-20DEC30-CDE`). */
  readonly product_id: string;
  /** Base-asset label (BTC/ETH); `null` without contracts enrichment. */
  readonly asset: string | null;
  readonly direction: DirectionLiteral;
  /** Signed net contracts held in this product (never 0 — flat rows are
   *  filtered server-side). */
  readonly contracts: number;
  /** Decimal-as-string. Quantity-weighted average entry price. */
  readonly entry_vwap: string | null;
  /** Decimal-as-string. Live WS mark; `null` when the mark store is absent
   *  or the mark exceeds the 3-minute staleness policy. */
  readonly mark: string | null;
  /** The worker's marks-stale flag (true — conservative — when the worker
   *  has never reported). */
  readonly marks_stale: boolean;
  /** Decimal-as-string. Client-side 2×ATR soft stop (risk-loop enforced). */
  readonly client_stop_price: string | null;
  /** Decimal-as-string fraction, SERVER-computed: (mark−stop)/mark for a
   *  long, (stop−mark)/mark for a short. `null` when mark or stop missing. */
  readonly client_stop_headroom_frac: string | null;
  /** Decimal-as-string. Venue-resting 3×ATR stop-limit backstop price. */
  readonly native_stop_price: string | null;
  readonly native_stop_headroom_frac: string | null;
  /** True when the worker tracks a venue-resting native stop order id.
   *  False with non-zero contracts = UNPROTECTED (see `isUnprotected`). */
  readonly native_stop_resting: boolean;
  /** True when the engine's stop fired on today's UTC bar-day. */
  readonly stopped_today: boolean;
}

export interface ExitProximityResponse {
  readonly positions: readonly PositionStopProximityView[];
  readonly server_now: string;
}

// ---------------------------------------------------------------------------
// Pure helpers — semantic predicates + the inverted state→token map.
// Extracted so the meaning lives in one place and the component stays a
// thin renderer. (No client-side sort — see the file header.)
// ---------------------------------------------------------------------------

/** Row-level stop status. INVERTED valence vs the entry side — see the
 *  file header. `holding` = protected, `near` = attention (marks stale),
 *  `triggered` = stop fired today. */
export type ExitStateLiteral = 'holding' | 'near' | 'triggered';

/**
 * Tailwind chip classes per exit state — the INVERTED-valence map.
 * HOLDING = green ("safe"), NEAR = yellow ("attention"), TRIGGERED = red.
 * Deliberately NOT shared with the entry-side chip map, whose green means
 * the opposite ("firing").
 */
export function exitStateChipClass(state: ExitStateLiteral): string {
  switch (state) {
    case 'holding':
      return 'bg-health-green/20 text-health-green';
    case 'near':
      return 'bg-health-yellow/20 text-health-yellow';
    case 'triggered':
      return 'bg-health-red/20 text-health-red';
  }
}

/**
 * A held position with NO native stop resting at the venue. The client
 * 2×ATR stop may still be armed (risk-loop enforced) but the venue-side
 * backstop is missing — a latent POSITION_UNPROTECTED risk (delta spec
 * §3.1 requires the 3×ATR backstop within 10 s of any position-opening
 * fill), so the row renders alert styling, not a calm green chip.
 */
export function isUnprotected(p: PositionStopProximityView): boolean {
  return p.native_stop_resting === false && p.contracts !== 0;
}

/**
 * Row status without inventing client-side thresholds: `triggered` when
 * the engine's stop fired today, `near` when the mark feed is stale
 * (headroom can't be trusted), else `holding`. The unprotected case is
 * surfaced separately via `isUnprotected` (alert-row styling).
 */
export function stopRowStatus(p: PositionStopProximityView): ExitStateLiteral {
  if (p.stopped_today) return 'triggered';
  if (p.marks_stale) return 'near';
  return 'holding';
}

// ---------------------------------------------------------------------------
// TanStack Query hook
// ---------------------------------------------------------------------------

const EXIT_PROXIMITY_QUERY_KEY = ['positions', 'exit-proximity'] as const;

export function useExitProximity(): UseQueryResult<ExitProximityResponse> {
  return useQuery({
    queryKey: EXIT_PROXIMITY_QUERY_KEY,
    queryFn: ({ signal }) =>
      apiCall<ExitProximityResponse>('/api/today/exit-proximity', { signal }),
    staleTime: 60_000,
  });
}

export const EXIT_PROXIMITY_KEY = EXIT_PROXIMITY_QUERY_KEY;
